### ISSUE TITLE
Dependencies are not picked from a `...-pkg.el` descriptor file anymore

### ISSUE THREAD

**Issue by doublep** [melpa/melpa#9285]
Example: https://melpa.org/#/ecukes, https://stable.melpa.org/#/ecukes

The package has file `ecukes-pkg.el` that declares 7 dependencies, but none is listed on those pages linked above or (more importantly) returned in `archive-contents`.

This seems to be a recent issue. Eldev's regression test suite depends on Ecukes being installable and last successful run was just two days ago: https://github.com/emacs-eldev/eldev/actions.

**Comment by marienz** [comment]
I found https://github.com/melpa/package-build/commit/760cf53114a6f27e86bc87b44726ced34bcda0ea which looks related / sounds like this was a deliberate change (note the second half of its commit message). https://github.com/melpa/melpa/commit/064927ca37912aba45f449635b42cd38c0e9d8cd might be the relevant MELPA change.

**Comment by tarsius** [comment]
For the time being Melpa gets information from `NAME-pkg.el` again, but I recommend that for the packages you maintain, you start to migrating the metadata to `NAME.el`.

**Comment by Fuco1** [comment]
Are the `-pkg.el` files somehow deprecated officially, or were there never really supported in the first place?

**Comment by tarsius** [comment]
I don't know whether it was expected that `NAME-pkg.el` would be used as a source of metadata when `package.el` was first created. I don't think the (largely unused) package creation functionality in `package-x.el` does use that file like so (but one could study the source to find out). `elpa-admin.el`, which is used for [Non]GNU Elpa, never used that file as a source.

Melpa prefers to get the metadata from `NAME.el` too, but it could (and now can again, for the time being) fall back to also getting metadata from `NAME-pkg.el`, if it is missing from `NAME.el`. That fall back is what I am trying to deprecate now.

The main motivation for that is that I would like to bring better version strings to Melpa and that involves extracting the current release, as seen by upstream, from somewhere. I learned that many package authors provide the information in multiple locations, and that a very significant cohort of them fails miserably at keeping these values in sync. The main sources of version information are Git tags and `[Package-]Version` library headers. The inconsistency is already grave if we only consider these to places, when we add a third place -- the version string specified in `NAME-pkg.el` -- it only gets worse.

It's not such that *many* `NAME-pkg.el` contain outdated information, the *majority* does. So support for these data corpses has to be phased out, but for this step in that direction (of many) I pulled the plug a bit to quickly.

**Comment by doublep** [comment]
> I pulled the plug a bit to quickly

Just to note, directly for my packages (I was the original reporter) this is not relevant, they don't use `...-pkg.el` themselves. However, they depend on certain packages (that I don't maintain) being installable from MELPA, because those are "standard specially supported tools" for Eldev and working integration with them is required as part of regression tests. And it's not only about _my_ regression tests, if they get broken, some users of Eldev can also be affected.

In other words, some packages are installed in an automated way, not only by someone's explicit command in Emacs package manager. And it's not nice if that gets broken somehow.

**Comment by marienz** [comment]
> I don't think the (largely unused) package creation functionality in package-x.el does use that file like so (but one could study the source to find out).

It looks like `package-upload-file` in package-x.el only supports uploading a previously-built tar file containing a multi-file package or a single-file package, not a directory. When handed a tar file, it uses `package-tar-file-info` from package.el, which uses NAME-pkg.el.

package.el also contains `package-dir-info` for getting package metadata from a directory, which tries to use NAME-pkg.el and falls back to NAME.el (so the opposite fallback from what Melpa does). That is used by `package-install-from-buffer` to install a package from a directory (as mentioned in the docstring of that function: "Specially, if current buffer is a directory, the -pkg.el description file is not mandatory, in which case the information is derived from the main .el file in the directory."). `package-dir-info` seems to have no other users within Emacs itself.

Although I may well have missed something, at first glance Emacs itself does not provide functionality for creating tar files suitable to be used/installed as packages. The Elisp manual describes "Multi-file packages" as tar files that must contain NAME-pkg.el (consistent with the implementation), not directories, but does not describe how to conveniently create/build such tar files from directories. And I did not spot tar-file-writing functionality in package.el or package-x.el. It seems to leave this to `elpa-admin.el` (and Melpa, etc).

This seems like a confusing situation: a package author who lays out their directory the same way the ELisp manual says their tar file should look will end up with something `package-install-from-buffer` accepts but `elpa-admin.el` (and Melpa, after this change) do not. The "Packaging" documentation in the ELisp manual seemingly does not describe the directory layout `elpa-admin.el` expects.

(I'm interested in all this because I have some code that currently uses `package-dir-info` to extract dependency information from packages: I found out from this issue that that falls back between NAME.el and NAME-pkg.el in the opposite order from what Melpa is doing, which is likely an accident waiting to happen....)
